lib/Sema/TypeCheckConstraints.cpp

```diff
@@ -1131,7 +1138,49 @@ TypeExpr *PreCheckExpression::simplifyTypeExpr(Expr *E) {
                                         ResultTypeRepr);
     return new (TC.Context) TypeExpr(TypeLoc(NewTypeRepr, Type()));
   }
-
+  
+  // Fold 'P & Q' into a composition type
+  if (auto *binaryExpr = dyn_cast<BinaryExpr>(E))
+    if (auto d = dyn_cast<OverloadedDeclRefExpr>(binaryExpr->getFn()))
+      for (auto decl : d->getDecls()) {
```

Comment:
This could possibly happen if the stdlib declarations of `&` are not availiable; I will do some work to see what happens if that is the case and work around it.

Also we are not really doing overload resolution here, just checking whether the expression the parser mistook for operator calls is actually a type composition with `&`.